Ww cr kl

kristenleib opened the pull request:
In the setup.sh file I changed it to "LCG_105c" and I added cutlfow and flavor scores to the WW_CR analyses skeleton.

kristenleib commented:
**Apologies I realized that I have forgot to add the comments each time I commit changes:**

 In the 2nd to last commit:
- Added selection criteria to only select events with 1 c-tagged jet, 1 s-tagged jet, and 2 light flavor tagged jets.

In the most recent commit:
- Calculates the invariant mass of W1 and W2 and prints out the values
- I added the framework to save all histograms to test.root (this is commented out in the version that I pushed)

kristenleib commented:
Calculated the mass of W1 and W2 using the truth jets. That still gave a large low mass tail and a high mean, so I redid the calculation using the chi^2 method. This reduced the tail, but the mean is now around ~91 GeV for both W1 and W2.

izamiza10 commented on PostCutCode/Root/AnalysisWWCR.cxx:
```diff
@@ -30,26 +48,536 @@ void AnalysisWWCR::run()
+        // double chi2_option2 = 
+        //     (pow(W1_option2.M() - m_W_true, 2) + pow(W2_option2.M() - m_W_true, 2)) / (sigma_W * sigma_W);
+
+        if (chi2_option1 <= chi2_option2) {
```
you need to find the min chi**2 of all 3 options right

kristenleib commented:
Apologies for a bit of a messy code. I did my first attempt at applying the truth daughters and made the pre-selection cut based on the decay channel (only selecting hadronic decays).

izamiza10 commented on PostCutCode/Root/AnalysisWWCR.cxx:
```diff
@@ -30,26 +41,907 @@ void AnalysisWWCR::run()
+
         if(i % 10000 == 0) std::cout<<"Done i: "<<i<<" out of "<<nEntries<<std::endl;
 
+        ROOT::VecOps::RVec<int>* Wm_jet_truth = truth_Wm_Daugthers_pdg();
```
Put outside event loop, ie after l76 for example

izamiza10 commented on PostCutCode/Root/AnalysisWWCR.cxx:
```diff
@@ -30,26 +41,907 @@ void AnalysisWWCR::run()
+        for (size_t i = 0; i < Wm_jet_truth->size(); ++i) {
+            int pdg_code = Wm_jet_truth->at(i);
+
+            if (pdg_code == 4) nTruth_c_W1++;
```
you can take the abs value & then you do not need to check if it is a particle or anti-particle

izamiza10 replied:
also from charge conservation you can tell which decay is allowed, ie W- -> u bar d/s bar c, it cannot go to u bar d or bar s c

izamiza10 commented on PostCutCode/Root/AnalysisWWCR.cxx:
```diff
@@ -30,26 +41,907 @@ void AnalysisWWCR::run()
+        int nTruth_c_W2 = 0, nTruth_s_W2 = 0, nTruth_u_W2 = 0, nTruth_d_W2 = 0;
+        int nTruth_cbar_W2 = 0, nTruth_sbar_W2 = 0, nTruth_ubar_W2 = 0, nTruth_dbar_W2 = 0;
+
+        for (size_t i = 0; i < Wp_jet_truth->size(); ++i) {
```
you can see that you are repeating the same code as above here for W-, which hints that it might be easier to write a function that does the same thing.

izamiza10 replied:
@kristenleib can you see my comment?

kristenleib commented:
- Fixed issues with the W1 and W2 mass
- Added a pre-selection cut on the W decays using the truth jets (only selecting hadronic decays)
- Converted this pre-selection cut into a function
- Calculated and plotted the momenta for W1->cs and W2->ud
